packages/ejson/lib/src/decoding.dart

```diff
@@ -32,43 +33,57 @@ const _commonDecoders = {
   Symbol: _decodeSymbol,
   Uint8List: _decodeBinary,
   Uuid: _decodeUuid,
+  DBRef: _decodeDBRef,
   Undefined: _decodeUndefined,
   UndefinedOr: _decodeUndefinedOr,
 };
 
-/// Custom decoders for specific types. Use `register` to add a custom decoder.
-final customDecoders = <Type, Function>{};
-
-final _decoders = () {
+/// Predefined decoders for common types
+final commonDecoders = () {
   // register extra common types on first access
   undefinedOr<T>(dynamic f) => f<UndefinedOr<T>>();
   TypePlus.addFactory(undefinedOr);
   TypePlus.addFactory(<T>(dynamic f) => f<Defined<T>>(), superTypes: [undefinedOr]);
   TypePlus.addFactory(<T>(dynamic f) => f<Undefined<T>>(), superTypes: [undefinedOr]);
+  TypePlus.addFactory(<T>(dynamic f) => f<DBRef<T>>());
   TypePlus.add<BsonKey>();
   TypePlus.add<ObjectId>();
   TypePlus.add<Uint8List>();
   TypePlus.add<Uuid>();
 
-  return CombinedMapView([customDecoders, _commonDecoders]);
+  return _commonDecoders;
 }();
 
+/// Custom decoders for specific types. Use `register` to add a custom decoder.
+final customDecoders = <Type, Function>{};
+
+final _decoders = CombinedMapView([customDecoders, commonDecoders]);
+
 /// Converts [ejson] to type [T].
 ///
+/// [defaultValue] is returned if set, and [ejson] is `null`.
+///
 /// Throws [InvalidEJson] if [ejson] is not valid for [T].
 /// Throws [MissingDecoder] if no decoder is registered for [T].
-T fromEJson<T>(EJsonValue ejson) {
-  final type = T;
-  final nullable = type.isNullable;
-  final decoder = nullable ? _decodeNullable : _decoders[type.base];
-  if (decoder == null) {
-    throw MissingDecoder._(ejson, type);
-  }
-  final args = nullable ? [type.nonNull] : type.args;
-  if (args.isEmpty) {
-    return decoder(ejson) as T; // minor optimization
+T fromEJson<T>(EJsonValue ejson, {bool? allowCustom, T? defaultValue}) {
```

Comment:
For `_allowCustom` think thread local storage, but since dart don't have shared heaps, we can just use a static variable. If the `allowCustom` parameter is set, it will use that for `_allowCustom` on any recursive calls to `fromEJson` such as would happen for list etc.

The problem I'm trying to solve is that you could `register` a type `T` with all optional values, meaning that `{}` would deserialize to `T` (and any other map for that matter, as redundant properties are already allowed). We don't try to search for best matching deserializer, so in the dynamic case, where the type to deserialize to is unknown (such as is the case for `RealmValue`), the custom deserializers gets in the way.

Hence this flag. I deliberately didn't document it, as it is very internal thing only really meant for `RealmValue`.

However, I will add, and elaborate the in-code comments.

Reply:
The:
```dart
      {'a': 1, 'b': 2}: {
        // Even though Game could be deserialized from this (everything in Game
        // is optional, hence {} and by extension any map can deserialize to Game),
        // we always decode a map as a map inside a RealmValue.
        'a': {'\$numberInt': '1'},
        'b': {'\$numberInt': '2'},
      },
```
test case illustrate the issue.

Reply:
And the following case in `_decodeAny`:
```dart
    Map<dynamic, dynamic> _ => _tryDecodeCustomIfAllowed(ejson) ?? _decodeDocument<String, dynamic>(ejson), // other maps goes last!!
```
Obviously any map can decode as a map, so we need to check custom decoders first, but in some cases that is not what you want either, hence the flag.

One could perhaps argue that `_decodeAny` should never even try the custom decoders, and always return a map for non BSOM types. Would yor prefer that?

Reply:
I have been thinking about it since yesterday. How do you feel about never trying the custom decoders in `_decodeAny`. Personally I think I prefer that for a couple of reasons:

1. Custom decoders will only be invoked when the type to deserialize to is known. Calling `fromEJson<dynamic>(..)`  (or implicit `fromEJson(..)` will never try the _custom_ decoders registered by the user. This removes ay ambiguity.
2. It is slightly more performant, as `_tryDecodeCustomIfAllowed` tries each custom decoder in turn, until the first match.
3. I can get rid of the `allowCustom`/`_allowCustom` shenanigans.

Reply:
Sure, that also sounds viable